### PR TITLE
Add dual-stack upgrade and multinode tests

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
### Summary
Added tests for dual-stack during clustering and upgrading of the cluster. The tests for clustering are added as a part of TestCluster to save redundancy of code for similar operations. 
For upgrading, we deploy a dual stack enabled cluster and then refresh/upgrade it to a newer revision and check if it still supports dual stack.

### CODE REVIEW

File: tests/test-cluster.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 extraCNIEnv:
   IPv4_SUPPORT: true
   IPv4_CLUSTER_CIDR: 10.3.0.0/16
-  IPv4_SERVICE_CIDR: 10.153.183.0/24
```
Comment (ktsakalozos) [inline tests/test-cluster.py]: Why did you change that to 152? Is there a problem with 1.53?

Comment (sachinkumarsingh092) [reply tests/test-cluster.py]: It was causing problems when we enabled DNS addon. This was because the [DNS IP](https://github.com/canonical/microk8s-core-addons/blob/main/addons/dns/enable#L50) is not in the range of this cidr.

File: tests/test-cluster.py [hunk 2]
```diff
@@ -558,28 +573,63 @@ def test_no_cert_reissue_in_nodes(self):
             lock_files = vm.run("ls /var/snap/microk8s/current/var/lock/")
             assert "no-cert-reissue" in lock_files.decode()
 
+    @pytest.mark.skipif(
+        # If the host system does not have IPv6 support or is not configured for IPv6, it won't be able to create VMs with IPv6 connectivity.
+        backend != "lxc" or not is_ipv6_configured,
```
Comment (ktsakalozos) [inline tests/test-cluster.py:578]: Multipass also supports IPv6, why are we explicit to lxc here?

File: tests/test-cluster.py [hunk 3]
```diff
@@ -558,28 +573,63 @@ def test_no_cert_reissue_in_nodes(self):
...
+                    print(resp)
+                    break
+            except subprocess.CalledProcessError as e:
+                print("Error occurred during the request:", str(e))
+                raise
+            attempt -= 1
+            time.sleep(2)
+
 
 class TestDualStack(object):
```
Comment (ktsakalozos) [inline tests/test-cluster.py:624]: Do we need this TestDualStack anymore? The test was moved to TestCluster so why do keep this class?

File: tests/test-cluster.py (hunk 2 above)
Comment (sachinkumarsingh092) [reply tests/test-cluster.py:578]: Due to the use of commands like these (https://github.com/canonical/microk8s/pull/4079/files#diff-8266c515875225e782094da6227c5ca720828f786e9f4c793b12e025e88f2e7aR585) which assume a lxc container for tests.

File: tests/test-cluster.py [hunk 4]
```diff
@@ -259,14 +274,14 @@ def setup_cluster(self):
                 for i in range(0, size):
                     print("Creating machine {}".format(i))
                     vm = VM(backend)
-                    vm.setup(channel_to_test)
```
Comment (ktsakalozos) [inline tests/test-cluster.py:262]: This is the place where you should check for IPv6 support and not inside the VM class.

File: tests/test-cluster.py [hunk 5]
```diff
@@ -259,14 +274,14 @@ def setup_cluster(self):
...
+                    vm.setup(channel_to_test, launch_config)
                     print("Waiting for machine {}".format(i))
                     vm.run("/snap/bin/microk8s.status --wait-ready --timeout 120")
                     self.VM.append(vm)
             else:
                 for vm_name in reuse_vms:
                     vm = VM(backend, vm_name)
-                    vm.setup(channel_to_test)
+                    vm.setup(channel_to_test, launch_config)
```
Comment (ktsakalozos) [inline tests/test-cluster.py:284]: Same as above.

File: tests/test-cluster.py [hunk 6]
```diff
@@ -24,6 +24,17 @@
 channel_to_test = os.environ.get("CHANNEL_TO_TEST", "latest/stable")
 backend = os.environ.get("BACKEND", None)
 profile = os.environ.get("LXC_PROFILE", "lxc/microk8s.profile")
+launch_config = """---
+version: 0.1.0
+extraCNIEnv:
+  IPv4_SUPPORT: true
+  IPv4_CLUSTER_CIDR: 10.3.0.0/16
+  IPv4_SERVICE_CIDR: 10.152.183.0/24
```
Comment (ktsakalozos) [inline tests/test-cluster.py:32]: Why we changed to 10.152.183.0 from 10.153.183.0?

File: tests/test-cluster.py [hunk 7]
```diff
@@ -87,7 +98,9 @@ def _setup_lxc(self, channel_or_snap, launch_configs=None):
                 ).split()
             )
             time.sleep(20)
-            self._load_launch_configuration_lxc(launch_configs)
+            if is_ipv6_configured:
```
Comment (ktsakalozos) [inline tests/test-cluster.py:101]: Do not do this here. Do the ipv6 check when you call the setup() method

File: tests/test-cluster.py [hunk 8]
```diff
@@ -87,7 +98,9 @@ def _setup_lxc(self, channel_or_snap, launch_configs=None):
...
+                self._load_launch_configuration_lxc(launch_configs)
```
Comment (ktsakalozos) [inline tests/test-cluster.py:102]: I also need to point out a blind spot you have. There is a second backend, multipass.

File: tests/test-cluster.py (hunk 6 above)
Comment (sachinkumarsingh092) [reply tests/test-cluster.py:32]: Same as above (https://github.com/canonical/microk8s/pull/4079#discussion_r1263715249)

File: tests/test-upgrade.py [hunk 9]
```diff
@@ -106,6 +108,32 @@ def test_upgrade(self):
             except CalledProcessError:
                 print("Will not test the metallb addon")
 
+        if is_ipv6_configured:
+            launch_config = """---
+version: 0.1.0
+extraCNIEnv:
+  IPv4_SUPPORT: true
+  IPv4_CLUSTER_CIDR: 10.3.0.0/16
+  IPv4_SERVICE_CIDR: 10.152.183.0/24
```
Comment (ktsakalozos) [inline tests/test-upgrade.py:117]: please change this to 10.153.183.0 so it matches the extraSANs

File: tests/test-upgrade.py [hunk 10]
```diff
@@ -106,6 +108,32 @@ def test_upgrade(self):
             except CalledProcessError:
                 print("Will not test the metallb addon")
 
+        if is_ipv6_configured:
```
Comment (ktsakalozos) [inline tests/test-upgrade.py:111]: Applying of the launch configuration needs to go when we first install the snap so we make sure it survives the refresh. If you put it here it will get ignored by the refresh.

File: tests/test-cluster.py (hunk 8 above)
Comment (sachinkumarsingh092) [reply tests/test-cluster.py:102]: Yeah but that is never triggered: https://github.com/canonical/microk8s/blob/6f488f61794fdd0ae30f626c8810acbfd0274768/tests/test-cluster.py#L188C9-L188C45. I'll fix this too

File: tests/test-cluster.py (hunk 2 above)
Comment (ktsakalozos) [reply tests/test-cluster.py:578]: Then we should implement this push in the VM class and we make the test below backend agnostic (as with the rest of the test) .

File: tests/test-cluster.py (hunk 3 above)
Comment (ktsakalozos) [reply tests/test-cluster.py:624]: We can drop the single node tests if we have multinode ones.

File: tests/test-cluster.py (hunk 1 above)
Comment (ktsakalozos) [reply tests/test-cluster.py]: This is a good catch. Lets create a follow up card to make the dns addon get the subnet from cni-env.

File: .github/workflows/build-snap.yml [hunk 11]
```diff
@@ -125,6 +125,8 @@ jobs:
     steps:
       - name: Checking out repo
         uses: actions/checkout@v3.3.0
+      - name: Setup tmate session
```
Comment (ktsakalozos) [inline .github/workflows/build-snap.yml:128]: Do we need this? Maybe we can keep it for debugging sessions

File: tests/templates/ingress.yaml [hunk 12]
```diff
@@ -18,7 +18,7 @@ spec:
         app: microbot
     spec:
       containers:
-        - image: cdkbot/microbot-$ARCH
+        - image: cdkbot/microbot-amd64
```
Comment (ktsakalozos) [inline tests/templates/ingress.yaml:21]: Please do not commit this.

File: tests/test-cluster.py (hunk 6 above)
Comment (ktsakalozos) [reply tests/test-cluster.py:32]: Now that we updated the core addon the 10.153.183.0 should work. Lets Try this.

File: tests/test-cluster.py [hunk 13]
```diff
@@ -87,7 +98,9 @@ def _setup_lxc(self, channel_or_snap, launch_configs=None):
                 ).split()
             )
             time.sleep(20)
+
```
Comment (ktsakalozos) [inline tests/test-cluster.py:101]: Do we need these empty lines?

File: .github/workflows/build-snap.yml (hunk 11 above)
Comment (sachinkumarsingh092) [reply .github/workflows/build-snap.yml:128]: It blocks the pipeline, I'll comment it out.

File: tests/templates/ingress.yaml (hunk 12 above)
Comment (sachinkumarsingh092) [reply tests/templates/ingress.yaml:21]: It always happens at local execution. Removing it.

File: tests/test-cluster.py (hunk 13 above)
Comment (sachinkumarsingh092) [reply tests/test-cluster.py:101]: Makes the separation b/w different steps easily visible imo.

File: tests/test-cluster.py (hunk 6 above)
Comment (sachinkumarsingh092) [reply tests/test-cluster.py:32]: Yes trying that.

File: tests/test-cluster.py [hunk 14]
```diff
@@ -24,6 +24,17 @@
...
+version: 0.1.0
+extraCNIEnv:
+  IPv4_SUPPORT: true
+  IPv4_CLUSTER_CIDR: 10.3.0.0/16
+  IPv4_SERVICE_CIDR: 10.153.183.0/24
+  IPv6_SUPPORT: true
+  IPv6_CLUSTER_CIDR: fd02::/64
+  IPv6_SERVICE_CIDR: fd99::/108
+extraSANs:
+  - 10.152.183.1"""
```
Comment (ktsakalozos) [inline tests/test-cluster.py:37]: This will also need to change

Comment (sachinkumarsingh092) [reply tests/test-cluster.py:37]: `extraSANs` to 10.153.x.x?
